python/lsst/ci/prepare.py

```diff
@@ -643,6 +690,121 @@ async def pull_worker(queue):
+            return owner, repo
+        else:
+            return None
+
+    def get_github_prs(self, owner, repo):
+        """Get the list of PRs using the GitHub API."""
+        # Get token set in util.jenkinsWrapper
+        token = os.environ["GITHUB_TOKEN"]
+
+        url = f"https://api.github.com/repos/{owner}/{repo}/pulls"
+        headers = {"Accept": "application/vnd.github.v3+json", "Authorization": f"token {token}"}
+
+        response = requests.get(url, headers=headers)
+        if response.status_code == 200:
+            prs = response.json()
+            return prs
+        else:
+            print(f"Failed to get PRs for {owner}/{repo}: {response.status_code}")
+            return []
+
+    async def list_non_default_refs_prs(self):
+        """List all PRs for non-default git refs and matching head refs"""
+        pr_info_file = os.path.join(self.build_dir, "pr_info.json")
+
+        # Remove any cached pr file before start
+        if os.path.exists(pr_info_file):
+            os.remove(pr_info_file)
+
+        pr_info = []
+
+        for product_name, product in self.product_index.items():
+            assert product.ref is not None
+            if product.ref.name == models.DEFAULT_BRANCH_NAME:
+                continue
+
+            # Get the repository URL
+            repo_dir = os.path.join(self.build_dir, product_name)
+            git = Git(repo_dir)
+            origin_url = await git("config", "--get", "remote.origin.url")
+
+            # Parse the GitHub repo owner and name from the origin URL
+            repo_info = self.extract_github_repo_info(origin_url)
+
+            if not repo_info:
+                print(f"Could not parse GitHub repo info from URL: {origin_url}")
+                continue
+
+            owner, repo = repo_info
+
+            # Retrieve PRs from Github
+            prs = self.get_github_prs(owner, repo)
+
+            matching_pr = []
+            for pr in prs:
+                pr_number = pr["number"]
+                pr_title = pr["title"]
+                pr_head_ref = pr["head"]["ref"]
+                pr_sha = pr["head"]["sha"]
+
+                # Check if the PR's head ref matches the non-default git ref
+                if pr_head_ref == product.ref.name:
+                    matching_pr.append(
+                        {
+                            "product_name": product_name,
+                            "owner": owner,
+                            "repo": repo,
+                            "pr_number": pr_number,
+                            "title": pr_title,
+                            "sha": pr_sha,
+                        }
+                    )
+
+            if matching_pr:
+                print(f"Found {len(matching_pr)} matching PR(s) for {product_name}:")
+                for match in matching_pr:
+                    print(f"PR #{match['pr_number']}: {match['title']}")
+
+                pr_info.extend(matching_pr)
+
+        if pr_info:
```

Comment:
Switch the pr_info to a pydantic model which can handle the JSON round tripping for you. You do not want to be constructing lists of dicts and writing to json and then reading it back as lists of dicts, you want to define the data model explicitly.